distributed/worker.py

```diff
@@ -2164,11 +2172,17 @@ def ensure_communicating(self):
                         in_flight = True
                         continue
                     host = get_address_host(self.address)
-                    local = [w for w in workers if get_address_host(w) == host]
+                    workers_not_busy = [
+                        w for w in workers if w not in self.busy_workers
+                    ]
+                    local = [w for w in workers_not_busy if get_address_host(w) == host]
                     if local:
                         worker = random.choice(local)
-                    else:
+                    elif not workers_not_busy:
+                        self.busy_workers.difference_update(workers)
```

Comment:
IIUC 
1. as soon as a worker is flagged busy it is added to `busy_workers`
2. If a worker is in `busy_workers`, it will always be ignored
3. The only way to remove a worker from this "busy" state is if all workers of the inspected tasks are flagged busy and only then they are removed from the `busy_workers`.


Condition 3.) is *very* strict since it hides a worker completely until we are lucky enough to hit a task with only busy `who_has` entries.

Reply:
You're following -- I will point out that the 2nd unchecked TODO on this PR is resetting a peer's busy state when fetching a new key belonging to that peer, which will reduce condition 3's strictness. However, I didn't want to go ahead and bother with figuring out how to do that until the larger discussion re: the best approach to dealing with this was resolved.